### PR TITLE
Use LambdaMetafactory where possible for lambda creation.

### PR DESCRIPTION
**Pull request by retronym**
My design notes: https://gist.github.com/retronym/0178c212e4bacffed568

This is all locked behind:

```
-Ydelambdafy:method -Ybackend:GenBCode -target:jvm-1.8
```

We require scala-java8-compat to be present on the compilation and runtime classpath. This provides a set of functional interfaces.

These are automatically downloaded and added to `scala-library.jar` if the property `scala-java8-compat.package` is set during the Ant build. This option is only intended to facilitate testing, we don't plan to ship those classes in any Scala release.

We can bootstrap with this enabled:

```
% cat build.properties
starr.use.released=1
% ant all.clean
% ant -Dscala-java8-compat.package="" locker.done
% ant -Dscalac.args=\"-Ydelambdafy:method -target:jvm-1.8\" -Dscala-java8-compat.package="" clean build-opt
% ./build/quick/bin/scala -target:jvm-1.8 -Ydelambdafy:method -Ybackend:GenBCode -e 'println((() => "").getClass)'
class Main$$anon$1$$Lambda$1/495053715
```
## Limitations
- Under this mode, lambdas are not serializable. This will be remedied in a followup PR.
- The inliner needs to be updated with an understanding of `LambdaMetafactory`'s
  semantics in order to be able to perform closure elimination.
- We are not able to mixin an the implementation of `toString` in to functions. So rather than printing as `<function>`, we instead get a rather ugly toString of: `$$Lambda$1/914374969@10289886`. We would require an extension point in `LambdaMetafactory`, or ship our own `ScalaLambdaMetafactory` to do better here.

We cannot yet run the full test suite under this mode, but we are getting close.

```
SCALAC_OPTS="-target:jvm-1.8 -Ydelambdafy:method -Ybackend:GenBCode" ./test/partest --run --pos --neg --terse
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=128M; support was removed in 8.0
Selected 3889 tests drawn from 3 named test categories

........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
.....................

........................................................................
........................................................................
........................................................................
........................................................................
.........................................................
!!   1 - neg/t3234.scala                           [expected compilation failure]
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
.....................................

........................................................................
..................
!!    1 - run/delambdafy-specialized.scala          [output differs]
.
!!    2 - run/delambdafyLambdaClassNames            [non-zero exit code]
........................................................................
..............................
!!    3 - run/lambda-test.scala                     [output differs]     
// this was a canary that I added, expecting a failure if we used LambdaMetafactory
// It prints `(() = "").getClass`.
// 
// % diff /Users/jason/code/scala2/test/files/run/lambda-test-run.log /Users/jason/code/scala2/test/files/run/lambda-test.check--- empty
// +++ lambda-test-run.log
// @@ -1,0 +1,1 @@
// +class Test$$$Lambda$1/402115881

........................................................................
........................................................................
..........................................................
!!    4 - run/private-inline.scala                  [output differs]
........................................................................
........................................................................
...........................................
!!    5 - run/repl-javap-app.scala                  [output differs]
........................................................................
......
!!    6 - run/synchronized.scala                    [output differs]
........................................................................
..
!!    7 - run/t2106.scala                           [output differs]
.....................................................................
!!    8 - run/t3158.scala                           [output differs]
........................................................................
........................................................................
.........................................................
!!    9 - run/t5018.scala                           [non-zero exit code]
........................................................................
........................................................................
.........................................
!!   10 - run/t6102.scala                           [output differs]
........................................................................
........................................................................
........................................................................
.................................
!!   11 - run/t7582                                 [output differs]
..
!!   12 - run/t7582b                                [output differs]
!!   13 - run/t7584b.scala                          [non-zero exit code]
........................................................................
..............
!!   14 - run/t8601-closure-elim.scala              [non-zero exit code]
................................
!!   15 - run/t8960.scala                           [non-zero exit code]
........................................................................
......................................
```

Review by @adriaanm @lrytz

### CODE REVIEW

**Comment by retronym** [comment]
While many of these the failures listed in the PR description are benign, there is at least one problematic one I'm investigating:

```
% qscalac -Ydelambdafy:method -Ybackend:GenBCode test/files/run/t7584b.scala && qscala TEst
topic/indylambda-emit-indy /code/scala2
% qscalac -target:jvm-1.8 -Ydelambdafy:method -Ybackend:GenBCode test/files/run/t7584b.scala && qscala Test
java.lang.NullPointerException
    at Test$$$Lambda$1/1587487668.apply(Unknown Source)
    at Test$.Test$$$anonfun$1(t7584b.scala:2)
    at Test$$$Lambda$2/1848402763.apply(Unknown Source)
    at scala.compat.java8.JFunction1.apply$mcII$sp(JFunction1.java:29)
    at Test$.delayedEndpoint$Test$1(t7584b.scala:8)
    at Test$delayedInit$body.apply(t7584b.scala:1)
    at scala.Function0$class.apply$mcV$sp(Function0.scala:40)
    at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
```

**Comment by retronym** [comment]
That one minimizes to:

``` scala
object Test {
  def main(args: Array[String]): Unit = {
    val f: (Int, => Int) => Int = (x, y) => 0
    f(null.asInstanceOf[Int], 0)
  }
}
```

**Comment by retronym** [comment]
Okay, this could be a problem. We had assumed that `LambdaMetafactory`'s boxing and unboxing would be suitable for our use (modulo conversion from `void` to `BoxedUnit`). But Scala is more liberal with unboxing null references:

``` scala
scala> Int.unbox(null)
res1: Int = 0

scala> val i2i = (x: Int) => x + 1
i2i: Int => Int = <function1>

scala> i2i.asInstanceOf[AnyRef => Int]
res4: AnyRef => Int = <function1>

scala> res4(null)
res5: Int = 1
```

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala [hunk 1]
```diff
@@ -315,10 +326,14 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
       if (sym.isClass) env.keySet
       else specializedTypeVars(sym).intersect(env.keySet)
     )
+    specializedName(sym.name, tvars, env)
+  }
+
+  private def specializedName(name: Name, tvars: immutable.Set[Symbol], env: TypeEnv): TermName = {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:332]: What't the reason for this new definition? There seems to be only the one callsite above.

**Comment by lrytz** [comment]
looks very good!

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:332]: IT does seem unnecessary (albeit not harmful) now, perhaps I had another caller during an earlier iteration of the patch.

**Comment by retronym** [comment]
If we are unable to use the unboxing synthesized by LMF (which is just a virtual call to `{int,long,...}Value` of the boxed representation that isn't null safe), we might be forced to emit our lamgda target method without primitive types in the signature. Instead, it would perform primitive boxing/unboxing inside that method. We could use the same approach for function's dealing with value classes (this PR punts on that thorny issue and uses the old encoding.)

``` scala
val f = (s: String) => 42
// def $target(s: String): Integer = Int.box(42)
// val f = indy(LambdaMetaFactory, <$target>)
```

**Comment by lrytz** [comment]
Would we lose specialization of functions in this approach?

**Comment by retronym** [comment]
With a specialized function, the abstract method already deals in primitives, so `LMF` does not need to create bridges to accomodate a lambda target with primitives in the signature. It only comes into play in functions like `String => Int`, which is not specialized.

**Comment by retronym** [comment]
A tradeoff is that we might incur unwanted boxing when the lambda target method is directly called from some Scala call site as a result of inlining, rather than being called through the generic interfaces method. e.g.

``` scala
def inlineMe = ((i: Int) => (i + 1).toString)(42)

def $lambdaTarget(i: Integer): String = Int.box((Int.unbox(i) + 1)).toString
def inlineMeOptimized = $lambdaTarget(Int.box(42))
```

In cases when we inline the lambda target as well, this peephole optimizations might get rid of the box/unbox pairs.

**Comment by lrytz** [comment]
OK, got it.

One idea for the inlining issue (might be too much overhead though): we could generate the lambda target methods as they are now, i.e. without boxing. Then we could have a second method that only boxes and delegates, and use this as the actual target for the metafactory.

This way, after inlining the lambda, the optimizer could rewrite `unbox(boxingLambdaTargetFoo)` directly to `lambdaTargetFoo`.

It would also make the workaround a bit more self-contained - there might be multiple places in the lambda target where boxing / unboxing has to be introduced (multiple return points).

**Comment by retronym** [comment]
To put this into context, I believe you need to cast to notice the NPE-on-unboxing issue.

So as long as we think it is solvable, we could tackle it in a followup. E.g. in the worst case, a clumsy but correct solution would be to use the old translation for non-specialized function types that deal in primitive value types.

In the meantime, it would be good to establish a foothold to start work on the inliner and serialization

**Comment by retronym** [comment]
Indirection through another method is certainly possible. Would be similar going through the static accessor (which I used to use for argument permutation.) We could teach the inliner to always inline that method, as I think it would always be beneficial. OTOH, we should seek to have the most direct call path when possible, so we might use this only in problematic cases.

**Comment by lrytz** [comment]
What is the situation with the void / BoxedUnit issue?

**Comment by adriaanm** [comment]
Super! Could you distill the boxing snags encountered and resolved in the design doc? Maybe illuminate our current choice and future refinements in terms of an overview of possible solutions?

**Comment by smarter** [comment]
FWIW, here's how I dealt with LMF and value classes in Dotty: https://github.com/smarter/dotty/commit/48af91130ddd6f15ebfd831617557625d8ff3bce (part of https://github.com/lampepfl/dotty/pull/411, Dotty always uses LMF for closures, so I had to implement something). This generates a "bridge" method which does the boxing/unboxing and calls the method which actually implements the closure.

Also, good catch with that null unboxing issue, supporting this is going to be annoying.

**Comment by retronym** [comment]
@lrytz Currently we solve the `void` -> `BoxedUnit` boxing by doing it ourselves in `JProcedureN` in scala-java8-compat.

```
scala> :javap -c scala.compat.java8.JProcedure0
Compiled from "JProcedure0.java"
public interface scala.compat.java8.JProcedure0 extends scala.compat.java8.JFunction0<scala.runtime.BoxedUnit> {
  public void $init$();
    Code:
       0: return

  public abstract void applyVoid();

  public scala.runtime.BoxedUnit apply();
    Code:
       0: aload_0
       1: invokeinterface #1,  1            // InterfaceMethod applyVoid:()V
       6: getstatic     #2                  // Field scala/runtime/BoxedUnit.UNIT:Lscala/runtime/BoxedUnit;
       9: areturn

  public java.lang.Object apply();
    Code:
       0: aload_0
       1: invokeinterface #3,  1            // InterfaceMethod apply:()Lscala/runtime/BoxedUnit;
       6: areturn
}
```

This was feasible as it doesn't add an explosion of functional interfaces: we just need an additional 22.

However, we also could solve this by pushing this into the lambda target method, as proposed above for numeric box/unboxing.

`JProcedure` might still be useful for folks creating `Function[_, Unit]` from Java source code; we don't want them to have to return BoxedUnit.INSTANCE.

**Comment by retronym** [comment]
@adriaanm https://gist.github.com/retronym/0178c212e4bacffed568#bridges-and-boxes

**Comment by adriaanm** [comment]
:sparkles:

**Comment by retronym** [comment]
Here's a spike at the DIY boxing approach: https://github.com/retronym/scala/compare/topic/indylambda-emit-indy...retronym:topic/indylambda-diy-boxing

**Comment by lrytz** [comment]
that can go into a subsequent PR, no?

**Comment by retronym** [comment]
Yes, I think it would add too much bulk to this one to try to include that.

**Comment by lrytz** [comment]
It's a bit unfortunate that we're not yet building on jdk 8, so we can't really have tests yet. Once a change is merged, the motivation to write tests for it usually declines :-)

**Comment by retronym** [comment]
I guess I ought to write the tests but leave them unmerged until we land on 2.12.x. Or we could make them noops on Java 6/7 but assert something on Java 8...

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala [hunk 2]
```diff
@@ -632,6 +633,10 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
             case _ =>
               abort(s"Cannot instantiate $tpt of kind: $generatedType")
           }
+        case Apply(_, args) if app.hasAttachment[delambdafy.LambdaMetaFactoryCapable] =>
+          val attachment = app.attachments.get[delambdafy.LambdaMetaFactoryCapable].get
+          genLoadArguments(args, paramTKs(app))
```
Comment (DarkDimius) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:638]: If `target` is non-static you should also load the receiver reference. I do not see where you do it.

Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:638]: In that case, `Delambdafy` rewrites to `$anonFunFactory(receiver, arg0, ... argN)`, where `$anonFunFactory` is a fictional symbol with a result type `(LambdaParam0, .. LambdaParam0) => LambdaResult`

**Comment by lrytz** [comment]
LGTM!

**Pull request merged**